### ISSUE TITLE
Intermittent failure with "'NoneType' object has no attribute '__getitem__'"

### ISSUE THREAD

**Issue by tj** [DataDog/docker-dd-agent#27]
Hey!

The agent seemed fine for a while, but I just noticed some weird patterns in our metrics, checked out the logs and it seems like the collector is failing with the errors below. Any suggestions? Seems like something upset it all of a sudden.

```
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.3.3
Git commit (client): 4749651/1.6.0
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.3.3
Git commit (server): 4749651/1.6.0
OS/Arch (server): linux/amd64
```

```
2015-05-29 14:08:12 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35090. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:11:23 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35100. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:14:33 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35110. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:17:44 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35120. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:20:54 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35130. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:24:06 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35140. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:27:16 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35150. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:30:27 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35160. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:33:37 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35170. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:36:48 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35180. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:39:59 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35190. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:43:10 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35200. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:46:20 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35210. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:49:31 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35220. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:52:41 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35230. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:55:53 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35240. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 14:59:03 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35250. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:02:14 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35260. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:05:24 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35270. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:08:35 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35280. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:11:46 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35290. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:14:56 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35300. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:18:07 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35310. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:21:17 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35320. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:24:29 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35330. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:27:39 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35340. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:30:50 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35350. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:34:00 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35360. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:37:11 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35370. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:40:22 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35380. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:43:33 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35390. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:46:43 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35400. Collection time: 4.03s. Emit time: 0.0s
2015-05-29 15:49:54 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35410. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:53:04 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35420. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:56:16 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35430. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 15:59:26 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35440. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:02:37 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35450. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:05:47 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35460. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:08:58 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35470. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:12:09 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35480. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:15:20 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35490. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:18:30 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35500. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:21:41 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35510. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:24:56 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 170, in _get_and_count_containers
    raise Exception("Failed to collect the list of containers. Exception: {0}".format(e))
Exception: Failed to collect the list of containers. Exception: timed out
2015-05-29 16:24:56 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35520. Collection time: 9.03s. Emit time: 0.01s
2015-05-29 16:25:21 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 170, in _get_and_count_containers
    raise Exception("Failed to collect the list of containers. Exception: {0}".format(e))
Exception: Failed to collect the list of containers. Exception: timed out
2015-05-29 16:28:13 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35530. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:31:23 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35540. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:34:34 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35550. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:37:44 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35560. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:40:56 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35570. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:44:06 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35580. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:47:17 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35590. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:50:27 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35600. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:53:38 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35610. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:56:49 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35620. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 16:59:27 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 170, in _get_and_count_containers
    raise Exception("Failed to collect the list of containers. Exception: {0}".format(e))
Exception: Failed to collect the list of containers. Exception: timed out
2015-05-29 17:00:05 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35630. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:01:45 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 170, in _get_and_count_containers
    raise Exception("Failed to collect the list of containers. Exception: {0}".format(e))
Exception: Failed to collect the list of containers. Exception: timed out
2015-05-29 17:02:04 UTC | INFO | dd.collector | aggregator(aggregator.py:358) | Metric docker.cpu.user has a rate < 0. Counter may have been Reset.
2015-05-29 17:03:20 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35640. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:06:31 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35650. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:09:41 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35660. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:12:53 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35670. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:16:03 UTC | INFO | dd.collector | checks.collector(collector.py:383) | Finished run #35680. Collection time: 4.03s. Emit time: 0.01s
2015-05-29 17:18:36 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 189, in _get_and_count_containers
    ids_to_names[container['Id']] = container['Names'][0].lstrip("/")
TypeError: 'NoneType' object has no attribute '__getitem__'
2015-05-29 17:18:55 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 547, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 135, in check
    containers, ids_to_names = self._get_and_count_containers(instance)
  File "/opt/datadog-agent/agent/checks.d/docker.py", line 189, in _get_and_count_containers
    ids_to_names[container['Id']] = container['Names'][0].lstrip("/")
TypeError: 'NoneType' object has no attribute '__getitem__'
2015-05-29 17:19:14 UTC | ERROR | dd.collector | checks.docker(__init__.py:556) | Check 'docker' instance #0 failed
Traceback (most recent call last):
````

**Comment by tj** [comment]
Image is 0281ae98acd21e438e65457afde5a3fefb2d896118a8a7a1611895f0480c686f, going to take a look at trying the others

**Comment by tj** [comment]
Same issue with the 5.3.0

**Comment by LotharSee** [comment]
Thanks for the detailed report!
It seems to be the same bug as the one we fixed here: https://github.com/DataDog/dd-agent/pull/1565

It is not released yet, but we provide a backported version: use the image `datadog/docker-dd-agent:backport`. Is that fixing it?

**Comment by tj** [comment]
Cool backport seems good so far, thanks!

**Comment by LotharSee** [comment]
Awesome! Thanks for your feedback.
As long as this is not part of the main branch `latest`, I will keep this issue open. This way you will know when to move back to `latest`!

**Comment by tj** [comment]
Hmm a few seemed to disappear again, not seeing any errors in the logs this time (yet)

**Comment by LotharSee** [comment]
It got fixed with the 5.4.0 release, latest and 5.4.x versions should fix it.
Also, I will remove the `backport` tag at some point. I will giving it a few extra weeks to live.

**Comment by PlugIN73** [comment]
@LotharSee got this issue with datadog-agent v5.5.1

2015-11-10 09:20:07 UTC | ERROR | dd.collector | checks.docker_daemon(__init__.py:689) | Check 'docker_daemon' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 672, in run
    self.check(copy.deepcopy(instance))
  File "/opt/datadog-agent/agent/checks.d/docker_daemon.py", line 221, in check
    containers_by_id = self._get_and_count_containers()
  File "/opt/datadog-agent/agent/checks.d/docker_daemon.py", line 277, in _get_and_count_containers
    container_name = str(container['Names'][0]).strip('/')
TypeError: 'NoneType' object has no attribute '__getitem__'
